### PR TITLE
handle discarding submodule changes properly

### LINKED ISSUE #3647: Discarding the revision of a submodule destroys the whole submodule
### Description

When you have a repository in a GitHub that contains a submodule and you discard the revision change of the submodule, the whole submodule is deleted.

### Version

**GitHub Desktop version:** 1.0.11

**OS version:** OS X 10.12.6

### Steps to Reproduce

1. Checkout a module that includes a submodule
1. Go to the submodule and change the branch
1. Observe the change in GitHub Desktop
1. Discard the change

**Expected behavior:** The submodule reference points back to the original branch and includes the files in that branch

**Actual behavior:** The whole submodule is completely destroyed. The folder is empty.

**Reproduces how often:** 100%

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #3647 

 - [x] don't move a submodule to the trash
 - [x] `git checkout -- [path]` doesn't work on submodules, gotta do `git submodule update -- [path]` to undo changes
 - [x] add tests against real repositories
   - [x] listing submodules
   - [x] update submodule to undo change

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
Given how big this is, it may take me a few hours to get this reviewed. I'll likely ping you with questions as they come up.

**Comment by shiftkey** [comment]
@iAmWillShepherd the `app/test/fixtures/submodule-basic-setup` folder (a new repository to aid with testing) is inflating the diff, but take your time

File: app/src/lib/git/submodule.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    //
+    // then the path to the submodule
+    //
+    // then the output of `git describe` for the submodule in braces
+    // we're not leveraging this in the app, so go and read the docs
+    // about it if you want to learn more:
+    //
+    // https://git-scm.com/docs/git-describe
+
+    const [path, describeOutput] = rest
```
Comment (iAmWillShepherd) [inline app/src/lib/git/submodule.ts]: I believe this line can be deleted and line 36 can be changed to `const [path, describeOutput] = entry.substr(42).split(/\s+/)`.

File: app/src/lib/stores/git-store.ts [hunk 2]
```diff
@@ -943,8 +945,13 @@ export class GitStore {
     const pathsToCheckout = new Array<string>()
     const pathsToReset = new Array<string>()
 
+    const submodules = await listSubmodules(this.repository)
+
     await queueWorkHigh(files, async file => {
-      if (file.status !== AppFileStatus.Deleted) {
+      const foundSubmodule = submodules.find(s => s.path === file.path)
```
Comment (iAmWillShepherd) [inline app/src/lib/stores/git-store.ts]: This can be changed to `const submoduleExists = submodules.some(s => s.path === file.path)` and line 352 deleted.

**Comment by shiftkey** [comment]
:apple:

**Comment by say25** [comment]
Don't mean to ruin this but I think this doesn't work anymore. I modified the gemoji library in Github Desktop and attempted to discard changes. It failed.

[2018-05-03.desktop.beta.log](https://github.com/desktop/desktop/files/1969747/2018-05-03.desktop.beta.log)

**Comment by shiftkey** [comment]
@say25 can you open a fresh issue with details to reproduce it? This error isn't once I've seen before:

```
2018-05-03T05:41:28.688Z - error: [ui] `git submodule status --` exited with an unexpected code: 139.
 fe965f16f6f690d40a738e75a46fad3f41f5daa3 app/static/common/choosealicense.com (fe965f1)
 690a433b668348c33087d34fafe2155ebe8bb4cf app/static/common/gitignore (remotes/origin/reduce-noise-1009-g690a433)

/Applications/GitHub Desktop.app/Contents/Resources/app/git/libexec/git-core/git-submodule: line 979: 12490 Segmentation fault: 11  git ${wt_prefix:+-C "$wt_prefix"} ${prefix:+--super-prefix "$prefix"} submodule--helper status ${GIT_QUIET:+--quiet} ${cached:+--cached} ${recursive:+--recursive} "$@"
``

**Pull request merged**